### DISCUSSION TITLE
Why doesn't `TryAsync` wrap up this exception?

### DISCUSSION THREAD

**Discussion by MrYossu** [louthy/language-ext#950] [Q&A] [answered]
I'm trying to use `TryAsync` to ensure that my methods return something the consumer can handle neatly. I thought that the point of `TryAsync` was that if an exception was thrown in the code inside, it would return the exception, which could be handled in the appropriate action passed to `Match`.

However, I have the following code. This makes calls to the Google Drive API, but the details of that aren't relevant...

```c#
    public TryAsync<DriveFile> GetParentFolder(string folderId) =>
      TryAsync(() => GetParentFolderDo(folderId));

    private async Task<DriveFile> GetParentFolderDo(string folderId) {
      FilesResource.GetRequest request = _service.Files.Get(folderId);
      request.Fields = "id, name, parents";
      DriveFile folder = await request.ExecuteAsync();
      request = _service.Files.Get(folder.Parents[0]);
      return await request.ExecuteAsync();
    }
```

This is fine when the `folderId` is not for the root folder. However, if I pass in the `Id` of the root folder, then the `folder.Parents` collection is `null` (why not empty? Don't ask me). This means that the penultimate line of code throws an NRE.

I was expecting the NRE to be caught and handled by `TryAsync`, but instead, my code crashes.

What am I doing wrong? Thanks

**Comment by Protiguous** [comment]
Don't mind me, I'm just a bystander on this project, but I don't think the issues you are having has anything to do with language-ext.
Your questions would be better asked somewhere like StackOverflow.com, in my opinion.
I would also suggest breaking down your lines so you're not trying to do everything in one function.

**Comment by MrYossu** [comment]
@Protiguous Thanks for the reply. Why don't you think this is to do with LanguageExt? As I understand it, the whole point of `Try` and `TryAsync` is that they catch exceptions and return a monad that wraps either the desired result or the exception. That's not a C# feature, it's part of the monad that is in this package.

Please clarify if I misunderstood. Thanks again.

**Reply by Protiguous** [reply]
Can you step into the function, line by line, and see when the NRE happens? It may be before language-ext even has a chance.

**Comment by MrYossu** [comment]
@Protiguous Duh, just realised that the code I posted isn't even using the `TryAsync` stuff properly! I had posted an old version of the code.

I've updated the question, and now you can see that all the action happens inside the call to `TryAsync()`, so the NRE, which happens when it tries to access the null `Parents` property of the `folder` variable is in fact within the scope of the LanguageExt call.

Sorry about the confusion, does the correct code clarify? Thanks again.

**Comment by MrYossu** [comment, marked as answer]
@Protiguous Duh part II.

Now I tried again with my updated code, it works fine!

I think I need more sleep! Thanks for the replies.

**Reply by Protiguous** [reply]
Hah, I'm glad you fixed it. Good luck on future code!

**Comment by MrYossu** [comment]
@Protiguous Thanks - with this track record I'm going to need it!
